File: api/Enkl.Api/Controllers/PublicQueryController.cs
using Enkl.Api.Auth;
using Enkl.Api.Domain.Entities;
using Enkl.Api.Services;
using Microsoft.AspNetCore.Authorization;
using Microsoft.AspNetCore.Mvc;
using Microsoft.AspNetCore.RateLimiting;

namespace Enkl.Api.Controllers;

/// <summary>
/// The app's first public/3rd-party-facing API surface — deliberately namespaced and versioned
/// apart from the internal "api/..." routes every other controller uses, since those are only ever
/// consumed by this app's own frontend and can evolve freely, while this one is now a contract
/// external callers depend on directly.
///
/// [AllowAnonymous] because auth here is the org API key (ApiKeyAuthFilter), not a user JWT —
/// AddAuthentication()'s JwtBearer handler would otherwise run first and 401 a request that never
/// carried a JWT at all. See ApiKeyAuthFilter's own doc comment for the no-enumeration-oracle
/// contract every failure mode here follows.
/// </summary>
[ApiController]
[AllowAnonymous]
[EnableRateLimiting("public-query")]
[Route("api/public/v1/queries")]
public class PublicQueryController : ControllerBase
{
    private readonly PublicQueryExecutionService _execution;

    public PublicQueryController(PublicQueryExecutionService execution)
    {
        _execution = execution;
    }

    [HttpGet("{savedQueryId:guid}/results")]
    [TypeFilter(typeof(ApiKeyAuthFilter))]
    public async Task<IActionResult> GetResults(Guid savedQueryId, CancellationToken ct)
    {
        var query = (SavedQuery)HttpContext.Items[ApiKeyAuthFilter.SavedQueryItemKey]!;
        var result = await _execution.ExecuteAsync(query.ProjectId, query.Sql, ct);
        return Ok(new { rows = result.Rows, truncated = result.Truncated });
    }
}

File: php-api/src/Controllers/PublicQueryController.php
<?php

declare(strict_types=1);

namespace Enkl\Api\Controllers;

use Enkl\Api\Auth\ApiKeyAuthMiddleware;
use Enkl\Api\Services\PublicQueryExecutionService;
use Psr\Http\Message\ResponseInterface as Response;
use Psr\Http\Message\ServerRequestInterface as Request;

/**
 * Ported from Controllers/PublicQueryController.cs. The app's first public/3rd-party-facing API
 * surface — see that file's own note on the deliberate /api/public/v1/ namespace/versioning and the
 * [AllowAnonymous]-equivalent reasoning (auth here is ApiKeyAuthMiddleware, not a user JWT — this
 * route is registered in routes.php WITHOUT JwtAuthMiddleware's usual RequireAuthMiddleware pairing).
 */
final class PublicQueryController extends BaseController
{
    public function getResults(Request $request, Response $response): Response
    {
        $query = $request->getAttribute(ApiKeyAuthMiddleware::SAVED_QUERY_ATTRIBUTE);
        $result = (new PublicQueryExecutionService())->execute($query['ProjectId'], $query['Sql']);
        return $this->json($response, $result);
    }
}
